crawl-ref/source/timed-effects.cc

```diff
@@ -1258,3 +1259,104 @@ int speed_to_duration(int speed)
+    const string branch_name = branches[br].abbrevname;
+    return you.props[CLOCK_KEY_ROOT + branch_name].get_int();
+}
+
+static int& _zot_clock() {
+    return _zot_clock_for(you.where_are_you);
+}
+
+static bool _zot_clock_active_in(branch_type br)
+{
+    return br != BRANCH_ABYSS && !player_has_orb();
+}
+
+// Is the zot clock running, or is it paused or stopped altogether?
+bool zot_clock_active() {
+    return _zot_clock_active_in(you.where_are_you);
+}
+
+static bool _over_zot_threshold(branch_type br) {
+    return _zot_clock_for(br) >= MAX_ZOT_CLOCK - BEZOTTING_THRESHOLD;
+}
+
+// If the player was in the given branch, would they suffer penalties for
+// nearing the end of the zot clock?
+bool bezotted_in(branch_type br)
+{
+    return _zot_clock_active_in(br) && _over_zot_threshold(br);
+}
+
+// Is the player suffering penalties from nearing the end of the zot clock?
+bool bezotted() {
+    return bezotted_in(you.where_are_you);
+}
+
+// How many times should the player have been drained by Zot?
+int bezotting_level() {
+    if (!bezotted())
+        return 0;
+    const int MAX_ZOTS = 5;
+    const int TURNS_PER_ZOT = (MAX_ZOT_CLOCK - BEZOTTING_THRESHOLD) / MAX_ZOTS;
+    const int over_thresh = _zot_clock() - (MAX_ZOT_CLOCK - BEZOTTING_THRESHOLD);
+    return over_thresh / TURNS_PER_ZOT + 1;
+}
+
+// Decrease the zot clock when the player enters a new level.
+void decr_zot_clock() {
+    if (!zot_clock_active()) {
+        return;
+    }
+    if (bezotted()) {
+        mpr("As you enter the new level, Zot loses track of you.");
+    }
+    int &zot = _zot_clock();
+    zot = max(0, zot - ZOT_CLOCK_PER_FLOOR);
+}
+
+// Odds of the zot clock incrementing every aut, expressed as odds
+// out of 1000 (aka 10x a percent chance).
+static unsigned _zot_clock_odds() {
+    const int base_odds = 100; // 10% per aut, aka on average 1/turn
+    if (have_passive(passive_t::slow_zot)) {
+        // down to 6.7% at full piety, aka once every 1.5 turns. (only movement
+        // (is slowed, not all actions, so we shouldn't give double clock!)
+        return base_odds - div_rand_round(you.piety, 6);
+    }
+    return base_odds;
+}
+
+void incr_zot_clock() {
+    const int clock_incr = binomial(you.time_taken, _zot_clock_odds(), 1000);
+    const int old_lvl = bezotting_level();
+    _zot_clock() += clock_incr;
+    if (!bezotted()) return;
+
+    if (_zot_clock() >= MAX_ZOT_CLOCK)
+    {
+        mpr("Zot has found you!");
+        ouch(INSTANT_DEATH, KILLED_BY_ZOT);
+        return;
+    }
+
+    if (!old_lvl) {
+        mpr("You have lingered too long in familiar places. Zot approaches. Travel to new levels before it's too late!");
+        drain_player(150, true, true);
+    } else if (bezotting_level() > old_lvl) {
+        mpr("Zot draws near...");
+        drain_player(75, true, true);
+    }
+}
+
+void set_initial_zot_clock(bool start_deep) {
+    // Give Delver a pile of extra time to compensate for 'wasted' levels
```

Comment:
they still get the clock decr when the enter these levels, right (setting the `max(0,...)` aside)? Wondering if 3 levels worth extra is too much.